File: groundstation-backend/internal/models/remote_cockpit.go
package models

import (
	"time"

	"gorm.io/gorm"
)

type VideoCodec string

const (
	VideoCodecH264 VideoCodec = "h264"
	VideoCodecH265 VideoCodec = "h265"
)

type VideoResolution string

const (
	VideoRes480P   VideoResolution = "480x360"
	VideoRes640P   VideoResolution = "640x480"
	VideoRes960P   VideoResolution = "960x540"
	VideoRes720P   VideoResolution = "1280x720"
	VideoRes1080P  VideoResolution = "1920x1080"
)

type CockpitMode string

const (
	CockpitModeIdle         CockpitMode = "idle"
	CockpitModeConnecting   CockpitMode = "connecting"
	CockpitModeFlying       CockpitMode = "flying"
	CockpitModeMission      CockpitMode = "mission"
	CockpitModeEmergency    CockpitMode = "emergency"
	CockpitModeDisconnected CockpitMode = "disconnected"
)

type CockpitSession struct {
	ID               uint64         `gorm:"primaryKey;autoIncrement" json:"id"`
	SessionID        string         `gorm:"type:varchar(64);uniqueIndex;not null" json:"session_id"`
	UAVID            uint64         `gorm:"index;not null" json:"uav_id"`
	PilotID          uint64         `gorm:"index;not null" json:"pilot_id"`
	StartTime        time.Time      `gorm:"index;not null" json:"start_time"`
	EndTime          *time.Time     `json:"end_time,omitempty"`
	Mode             CockpitMode    `gorm:"type:varchar(20);default:'idle'" json:"mode"`
	TotalFlightMs    int64          `json:"total_flight_time_ms"`
	CommandsSent     int64          `json:"commands_sent"`
	FailoverEvents   int            `json:"failover_events"`
	AutoFallbackUsed bool           `json:"auto_fallback_used"`
	CreatedAt        time.Time      `json:"created_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`

	UAV   *UAV   `gorm:"foreignKey:UAVID" json:"uav,omitempty"`
	Pilot *User  `gorm:"foreignKey:PilotID" json:"pilot,omitempty"`
}

func (CockpitSession) TableName() string {
	return "cockpit_sessions"
}

type VideoStreamSession struct {
	ID               uint64         `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID            uint64         `gorm:"index;not null" json:"uav_id"`
	Codec            VideoCodec     `gorm:"type:varchar(10);default:'h265'" json:"codec"`
	Resolution       VideoResolution `gorm:"type:varchar(20);default:'1280x720'" json:"resolution"`
	TargetBitrateKbps int           `json:"target_bitrate_kbps"`
	CurrentBitrateKbps int          `json:"current_bitrate_kbps"`
	FPS              int            `json:"fps"`
	KeyframeInterval int            `json:"keyframe_interval"`
	AdaptiveEnabled  bool           `gorm:"default:true" json:"adaptive_enabled"`
	MinBitrateKbps   int            `json:"min_bitrate_kbps"`
	MaxBitrateKbps   int            `json:"max_bitrate_kbps"`
	MinResolution    VideoResolution `gorm:"type:varchar(20);default:'640x480'" json:"min_resolution"`
	MaxResolution    VideoResolution `gorm:"type:varchar(20);default:'1920x1080'" json:"max_resolution"`
	StreamURL        string         `gorm:"type:varchar(512)" json:"stream_url"`
	Protocol         string         `gorm:"type:varchar(20);default:'webrtc'" json:"protocol"`
	Active           bool           `gorm:"default:false" json:"active"`
	FramesDecoded    int64          `json:"frames_decoded"`
	FramesDropped    int64          `json:"frames_dropped"`
	LatencyMs        int            `json:"latency_ms"`
	JitterMs         int            `json:"jitter_ms"`
	PacketLoss       float64        `json:"packet_loss"`
	LastFrameTime    *time.Time     `json:"last_frame_time,omitempty"`
	StartedAt        *time.Time     `json:"started_at,omitempty"`
	StoppedAt        *time.Time     `json:"stopped_at,omitempty"`
	QualityAdjustments int          `json:"quality_adjustments"`
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`

	UAV *UAV `gorm:"foreignKey:UAVID" json:"uav,omitempty"`
}

func (VideoStreamSession) TableName() string {
	return "video_stream_sessions"
}

type CockpitLinkSnapshot struct {
	ID                   uint64         `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID                uint64         `gorm:"index;not null" json:"uav_id"`
	PrimaryLink          LinkType       `gorm:"not null;default:2" json:"primary_link"`
	SecondaryLink        LinkType       `gorm:"not null;default:1" json:"secondary_link"`
	PrimaryState         LinkState      `gorm:"not null;default:0" json:"primary_state"`
	SecondaryState       LinkState      `gorm:"not null;default:0" json:"secondary_state"`
	PrimaryLatencyMs     int            `json:"primary_latency_ms"`
	SecondaryLatencyMs   int            `json:"secondary_latency_ms"`
	PrimaryPacketLoss    float64        `json:"primary_packet_loss"`
	SecondaryPacketLoss  float64        `json:"secondary_packet_loss"`
	FailoverEnabled      bool           `gorm:"default:true" json:"failover_enabled"`
	FailoverThresholdMs  int            `gorm:"default:200" json:"failover_threshold_ms"`
	FailoverCount        int            `json:"failover_count"`
	LastFailoverTime     *time.Time     `json:"last_failover_time,omitempty"`
	AutoMissionFallback  bool           `gorm:"default:true" json:"auto_mission_fallback"`
	Timestamp            time.Time      `gorm:"index;not null" json:"timestamp"`
	CreatedAt            time.Time      `json:"created_at"`
	DeletedAt            gorm.DeletedAt `gorm:"index" json:"-"`
}

func (CockpitLinkSnapshot) TableName() string {
	return "cockpit_link_snapshots"
}

type NetworkMetricsLog struct {
	ID                  uint64         `gorm:"primaryKey;autoIncrement" json:"id"`
	UAVID               uint64         `gorm:"index;not null" json:"uav_id"`
	BandwidthKbps       float64        `json:"bandwidth_estimate_kbps"`
	RTTms               int            `json:"rtt_ms"`
	PacketLoss          float64        `json:"packet_loss"`
	JitterMs            int            `json:"jitter_ms"`
	ThroughputKbps      float64        `json:"throughput_kbps"`
	Timestamp           time.Time      `gorm:"index;not null" json:"timestamp"`
	CreatedAt           time.Time      `json:"created_at"`
	DeletedAt           gorm.DeletedAt `gorm:"index" json:"-"`
}

func (NetworkMetricsLog) TableName() string {
	return "network_metrics_logs"
}

type FlightControlLog struct {
	ID         uint64         `gorm:"primaryKey;autoIncrement" json:"id"`
	SessionID  string         `gorm:"type:varchar(64);index" json:"session_id"`
	UAVID      uint64         `gorm:"index;not null" json:"uav_id"`
	PilotID    uint64         `gorm:"index;not null" json:"pilot_id"`
	Pitch      float64        `json:"pitch"`
	Roll       float64        `json:"roll"`
	Yaw        float64        `json:"yaw"`
	Throttle   float64        `json:"throttle"`
	Source     string         `gorm:"type:varchar(20);default:'gamepad'" json:"source"`
	Timestamp  time.Time      `gorm:"index;not null" json:"timestamp"`
	CreatedAt  time.Time      `json:"created_at"`
	DeletedAt  gorm.DeletedAt `gorm:"index" json:"-"`
}

func (FlightControlLog) TableName() string {
	return "flight_control_logs"
}

File: groundstation-frontend/src/types/remote-cockpit.ts
import type { LinkType, LinkState } from './link'
import type { UAVMode } from './uav'

export enum VideoCodec {
  H264 = 'h264',
  H265 = 'h265'
}

export enum VideoResolution {
  RES_1920x1080 = '1920x1080',
  RES_1280x720 = '1280x720',
  RES_960x540 = '960x540',
  RES_640x480 = '640x480',
  RES_480x360 = '480x360'
}

export enum VideoQualityPreset {
  ULTRA_LOW = 'ultra_low',
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  ULTRA_HIGH = 'ultra_high'
}

export enum CockpitMode {
  IDLE = 'idle',
  CONNECTING = 'connecting',
  FLYING = 'flying',
  MISSION = 'mission',
  EMERGENCY = 'emergency',
  DISCONNECTED = 'disconnected'
}

export enum HIDDeviceType {
  GAMEPAD = 'gamepad',
  JOYSTICK = 'joystick',
  RC_TRANSMITTER = 'rc_transmitter',
  UNKNOWN = 'unknown'
}

export interface VideoStreamConfig {
  codec: VideoCodec
  resolution: VideoResolution
  bitrate_kbps: number
  fps: number
  keyframe_interval: number
  adaptive_enabled: boolean
  min_bitrate_kbps: number
  max_bitrate_kbps: number
  min_resolution: VideoResolution
  max_resolution: VideoResolution
}

export interface VideoStreamStatus {
  active: boolean
  codec: VideoCodec
  resolution: VideoResolution
  current_bitrate_kbps: number
  target_bitrate_kbps: number
  fps: number
  latency_ms: number
  jitter_ms: number
  packet_loss: number
  frames_decoded: number
  frames_dropped: number
  last_frame_time: number
}

export interface NetworkMetrics {
  bandwidth_estimate_kbps: number
  rtt_ms: number
  packet_loss: number
  jitter_ms: number
  throughput_kbps: number
  timestamp: number
}

export interface HIDAxisState {
  pitch: number
  roll: number
  yaw: number
  throttle: number
}

export interface HIDButtonState {
  arm: boolean
  disarm: boolean
  takeoff: boolean
  land: boolean
  rtl: boolean
  pause: boolean
  mode_switch: boolean
  emergency_stop: boolean
}

export interface HIDDeviceInfo {
  id: string
  name: string
  type: HIDDeviceType
  connected: boolean
  vendor_id: number
  product_id: number
  mapping_profile: string
  last_active: number
}

export interface HIDState {
  devices: HIDDeviceInfo[]
  active_device_id: string | null
  axes: HIDAxisState
  buttons: HIDButtonState
  enabled: boolean
  calibration_needed: boolean
}

export interface CockpitLinkStatus {
  primary_link: LinkType
  secondary_link: LinkType
  primary_state: LinkState
  secondary_state: LinkState
  primary_latency_ms: number
  secondary_latency_ms: number
  primary_packet_loss: number
  secondary_packet_loss: number
  failover_enabled: boolean
  failover_threshold_ms: number
  failover_count: number
  last_failover_time: number | null
}

export interface FlightControlCommand {
  uav_id: string
  pitch: number
  roll: number
  yaw: number
  throttle: number
  mode?: UAVMode
  timestamp: number
  source: 'keyboard' | 'gamepad' | 'autopilot'
}

export interface CockpitSession {
  session_id: string
  uav_id: string
  pilot_id: string
  start_time: number
  end_time: number | null
  mode: CockpitMode
  total_flight_time_ms: number
  commands_sent: number
  failover_events: number
}

export interface RemoteCockpitState {
  active: boolean
  mode: CockpitMode
  current_uav_id: string | null
  available_uav_ids: string[]
  video_config: VideoStreamConfig
  video_status: VideoStreamStatus
  network_metrics: NetworkMetrics | null
  network_metrics_history: NetworkMetrics[]
  link_status: CockpitLinkStatus | null
  hid: HIDState
  current_session: CockpitSession | null
  auto_mission_fallback: boolean
  last_video_disconnect_time: number | null
  quality_adjustment_count: number
  last_quality_adjustment_time: number | null
}

export const VideoQualityPresetConfig: Record<VideoQualityPreset, VideoStreamConfig> = {
  [VideoQualityPreset.ULTRA_LOW]: {
    codec: VideoCodec.H265,
    resolution: VideoResolution.RES_480x360,
    bitrate_kbps: 500,
    fps: 15,
    keyframe_interval: 30,
    adaptive_enabled: true,
    min_bitrate_kbps: 200,
    max_bitrate_kbps: 1000,
    min_resolution: VideoResolution.RES_480x360,
    max_resolution: VideoResolution.RES_640x480
  },
  [VideoQualityPreset.LOW]: {
    codec: VideoCodec.H265,
    resolution: VideoResolution.RES_640x480,
    bitrate_kbps: 1000,
    fps: 20,
    keyframe_interval: 40,
    adaptive_enabled: true,
    min_bitrate_kbps: 500,
    max_bitrate_kbps: 2000,
    min_resolution: VideoResolution.RES_480x360,
    max_resolution: VideoResolution.RES_960x540
  },
  [VideoQualityPreset.MEDIUM]: {
    codec: VideoCodec.H265,
    resolution: VideoResolution.RES_960x540,
    bitrate_kbps: 2500,
    fps: 25,
    keyframe_interval: 50,
    adaptive_enabled: true,
    min_bitrate_kbps: 1000,
    max_bitrate_kbps: 4000,
    min_resolution: VideoResolution.RES_640x480,
    max_resolution: VideoResolution.RES_1280x720
  },
  [VideoQualityPreset.HIGH]: {
    codec: VideoCodec.H265,
    resolution: VideoResolution.RES_1280x720,
    bitrate_kbps: 5000,
    fps: 30,
    keyframe_interval: 60,
    adaptive_enabled: true,
    min_bitrate_kbps: 2000,
    max_bitrate_kbps: 8000,
    min_resolution: VideoResolution.RES_960x540,
    max_resolution: VideoResolution.RES_1920x1080
  },
  [VideoQualityPreset.ULTRA_HIGH]: {
    codec: VideoCodec.H265,
    resolution: VideoResolution.RES_1920x1080,
    bitrate_kbps: 10000,
    fps: 60,
    keyframe_interval: 60,
    adaptive_enabled: true,
    min_bitrate_kbps: 5000,
    max_bitrate_kbps: 15000,
    min_resolution: VideoResolution.RES_1280x720,
    max_resolution: VideoResolution.RES_1920x1080
  }
}

export const ResolutionOrder: VideoResolution[] = [
  VideoResolution.RES_480x360,
  VideoResolution.RES_640x480,
  VideoResolution.RES_960x540,
  VideoResolution.RES_1280x720,
  VideoResolution.RES_1920x1080
]

export const getResolutionDimensions = (resolution: VideoResolution): { width: number; height: number } => {
  const [w, h] = resolution.split('x').map(Number)
  return { width: w, height: h }
}

export const CockpitModeText: Record<CockpitMode, string> = {
  [CockpitMode.IDLE]: '待机',
  [CockpitMode.CONNECTING]: '连接中',
  [CockpitMode.FLYING]: '飞行中',
  [CockpitMode.MISSION]: '航线飞行',
  [CockpitMode.EMERGENCY]: '紧急状态',
  [CockpitMode.DISCONNECTED]: '已断开'
}

export const VideoCodecText: Record<VideoCodec, string> = {
  [VideoCodec.H264]: 'H.264',
  [VideoCodec.H265]: 'H.265'
}

export const VideoQualityPresetText: Record<VideoQualityPreset, string> = {
  [VideoQualityPreset.ULTRA_LOW]: '超低画质',
  [VideoQualityPreset.LOW]: '低画质',
  [VideoQualityPreset.MEDIUM]: '中等画质',
  [VideoQualityPreset.HIGH]: '高画质',
  [VideoQualityPreset.ULTRA_HIGH]: '超高画质'
}

export const HIDDeviceTypeText: Record<HIDDeviceType, string> = {
  [HIDDeviceType.GAMEPAD]: '游戏手柄',
  [HIDDeviceType.JOYSTICK]: '飞行摇杆',
  [HIDDeviceType.RC_TRANSMITTER]: '遥控器',
  [HIDDeviceType.UNKNOWN]: '未知设备'
}
